crates/ty_python_semantic/resources/mdtest/narrow/conditionals/nested.md

```diff
@@ -275,7 +275,49 @@ def f(a: A):
     a.x = None
 ```
 
-Narrowing is invalidated if a `nonlocal` declaration is made within a lazy scope.
+The opposite is not true, that is, if a root expression is reassigned, narrowing on the member are
+no longer valid in the inner lazy scope.
+
+```py
+def f(l: list[str | None]):
+    if l[0] is not None:
+        def _():
+            reveal_type(l[0])  # revealed: str | None
+        l = [None]
+
+def f(l: list[str | None]):
+    l[0] = "a"
+    def _():
+        reveal_type(l[0])  # revealed: str | None
+    l = [None]
+
+def f(l: list[str | None]):
+    l[0] = "a"
+    def _():
+        l: list[str | None] = [None]
+        def _():
+            # TODO: should be `str | None`
+            reveal_type(l[0])  # revealed: Unknown
```

Comment:
The reason this is `Unknown` is because lazy snapshots only record bindings, not declarations.

Whether to include declaration information in the lazy snapshot will be an item for future consideration.

Reply:
I think we should infer the type `list[str | None]` for `l` in the binding above, this is just incomplete work in our generics implementation.